src/lightning/pytorch/callbacks/model_checkpoint.py

```diff
@@ -241,9 +249,10 @@ def __init__(
         self._last_global_step_saved = 0  # no need to save when no steps were taken
         self._last_time_checked: Optional[float] = None
         self.current_score: Optional[Tensor] = None
-        self.best_k_models: Dict[str, Tensor] = {}
+        self.best_k_models: Dict[str, Dict[str, Tensor | Dict[str, Tensor]]] = {}
```

Comment:
this may be easier to read
`Dict[str, Dict[str, Tensor]] | Dict[str, Dict[str, Dict[str, Tensor]]]`
 but ultimately we'd be better off defining a type alias

more importantly, we need to avoid breaking backward compatibility here
so whatever code relies on `best_k_models` being `Dict[str, Tensor]` today needs to keep working

I suggest we just limit ourselves to track `best_model_metrics` and not mess with `best_k_models`, or use a separate private attribute